framework/DataObjects/DataSet.py

```diff
@@ -0,0 +1,1853 @@
+    self._resetScaling()
+
+  def addVariable(self,varName,values,classify='meta'):
+    """
+      Adds a variable/column to the data.  "values" needs to be as long as self.size.
+      @ In, varName, str, name of new variable
+      @ In, values, np.array, new values (floats/str for scalars, xr.DataArray for hists)
+      @ In, classify, str, optional, either 'input', 'output', or 'meta'
+      @ Out, None
+    """
+    # TODO might be removable
+    assert(isinstance(values,np.ndarray))
+    assert(len(values) == self.size)
+    assert(classify in ['input','output','meta'])
+    # first, collapse existing entries
+    self.asDataset()
+    # format as single data array
+    # TODO worry about sampleTag values?
+    column = self._collapseNDtoDataArray(values,varName,labels=self._data[self.sampleTag])
+    # add to the dataset
+    self._data = self._data.assign(**{varName:column})
+    if classify == 'input':
+      self._inputs.append(varName)
+    elif classify == 'output':
+      self._outputs.append(varName)
+    else:
+      self._metavars.append(varName)
+    self._orderedVars.append(varName)
+
+  def asDataset(self, outType=None):
+    """
+      Casts this dataObject as dictionary or an xr.Dataset depending on outType.
+      @ In, outType, str, optional, type of output object (xr.Dataset or dictionary).
+      @ Out, xr.Dataset or dictionary.
+    """
+    if outType is None or outType=='xrDataset':
+      # return the xArray, i.e., the old asDataset()
+      return self._convertToXrDataset()
+    elif outType=='dict':
+      # return a dict
+      return self._convertToDict()
+    else:
+      # raise an error
+      self.raiseAnError(ValueError, 'DataObject method "asDataset" has been called with wrong '
+                                    'type: ' +str(outType) + '. Allowed values are: xrDataset, dict.')
+
+  def checkIndexAlignment(self,indexesToCheck=None):
+    """
+      Checks that all realizations share common coordinates along given indexes.
+      That is, assures data is not sparse, but full (no NaN entries).
+      @ In, indexesToCheck, list(str) or str or None, optional, indexes to check (or single index if string, or if None will check ALL indexes)
+      @ Out, same, bool, if True then alignment is good
+    """
+    # format request so that indexesToCheck is always a list
+    if mathUtils.isAString(indexesToCheck):
+      indexesToCheck = [indexesToCheck]
+    elif indexesToCheck is None:
+      indexesToCheck = self.indexes[:]
+    else:
+      try:
+        indexesToCheck = list(indexesToCheck) # TODO what if this errs?
+      except TypeError:
+        self.raiseAnError('Unrecognized input to checkIndexAlignment!  Expected list, string, or None, but got "{}"'.format(type(indexesToCheck)))
+    # check the alignment of each index by checking for NaN values in each slice
+    data = self.asDataset()
+    for index in indexesToCheck:
+      # check that index is indeed an index
+      assert(index in self.indexes)
```

Comment:
We've picked up using assertions when we're confident an error will not come because of a user action, but can only come from future development.  That way, when the user runs raven_framework, it runs in python opt mode and skips the assertion checks.

As a result, any assertions that CAN be failed by a user action should not be assertions.